File: ScorpioProto/Java/src/Scorpio/Conversion/DefaultReader.java
package Scorpio.Conversion;
import java.io.EOFException;
import java.io.IOException;
import java.io.InputStream;
import java.util.Date;
public class DefaultReader implements IReader {
	private byte readBuffer[] = new byte[8];
	private InputStream stream;
	private boolean closeStream;
	public DefaultReader(InputStream stream, boolean closeStream) {
		this.stream = stream;
		this.closeStream = closeStream;
	}
	private final void readFully(byte b[], int off, int len) throws IOException {
		int n = 0;
        while (n < len) {
            int count = stream.read(b, n, len - n);
            if (count < 0)
                throw new EOFException();
            n += count;
        }
    }
	public boolean ReadBool() throws Exception {
		return ReadInt8() == (byte) 1;
	}
	public byte ReadInt8() throws Exception {
		return (byte) stream.read();
	}
	public int ReadUInt8() throws Exception {
		return stream.read();
	}
	public short ReadInt16() throws Exception {
		return (short)ReadUInt16();
	}
	public int ReadUInt16() throws Exception {
		int ch1 = stream.read();
        int ch2 = stream.read();
        return (ch2 << 8) + (ch1 << 0);
	}
	public int ReadInt32() throws Exception {
		int ch1 = stream.read();
        int ch2 = stream.read();
        int ch3 = stream.read();
        int ch4 = stream.read();
		return ((ch4 << 24) + (ch3 << 16) + (ch2 << 8) + (ch1 << 0));
	}
	public int ReadUInt32() throws Exception {
		return ReadInt32();
	}
	public long ReadInt64() throws Exception {
		readFully(readBuffer, 0, 8);
        return (((long)readBuffer[7] << 56) +
                ((long)(readBuffer[6] & 255) << 48) +
                ((long)(readBuffer[5] & 255) << 40) +
                ((long)(readBuffer[4] & 255) << 32) +
                ((long)(readBuffer[3] & 255) << 24) +
                ((readBuffer[2] & 255) << 16) +
                ((readBuffer[1] & 255) <<  8) +
                ((readBuffer[0] & 255) <<  0));
	}
	public long ReadUInt64() throws Exception {
		return ReadInt64();
	}
	public float ReadFloat() throws Exception {
		return Float.intBitsToFloat(ReadInt32());
	}
	public double ReadDouble() throws Exception {
		return Double.longBitsToDouble(ReadInt64());
	}
	public String ReadString() throws Exception {
		int length = ReadUInt16();
		if (length == 0) { return ""; }
		byte[] bytes = new byte[length];
		readFully(bytes, 0, length);
		return new String(bytes, "utf-8");
	}
	public Date ReadDateTime() throws Exception {
		return new Date(ReadInt64());
	}
	public byte[] ReadBytes() throws Exception {
		int length = ReadInt32();
		byte[] bytes = new byte[length];
		readFully(bytes, 0, length);
		return bytes;
	}
	public void Close() throws Exception {
		if (closeStream) {
			this.stream.close();
		}
	}
}

File: Scorpio.Conversion.Runtime/CSharp/Scorpio.Conversion.Runtime/src/Scorpio.Conversion.Runtime/DefaultReader.cs
using System;
using System.IO;
using System.Text;
namespace Scorpio.Conversion.Runtime {
    public class DefaultReader : IReader {
        private readonly static DateTime BaseTime = new DateTime(1970, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc);
        BinaryReader reader;
        Stream stream;
        bool closeStream;
        public DefaultReader(string file) : this(File.OpenRead(file), true) { }
        public DefaultReader(byte[] buffer) : this(new MemoryStream(buffer), true) { }
        public DefaultReader(Stream stream, bool closeStream = false) {
            this.stream = stream;
            this.reader = new BinaryReader(stream);
            this.closeStream = closeStream;
        }
        public virtual void ReadHead() {
            ConversionUtil.ReadHead(this);
        }
        public virtual bool ReadBool() {
            return ReadInt8() == 1;
        }
        public virtual sbyte ReadInt8() {
            return reader.ReadSByte();
        }
        public virtual byte ReadUInt8() {
            return reader.ReadByte();
        }
        public virtual short ReadInt16() {
            return reader.ReadInt16();
        }
        public virtual ushort ReadUInt16() {
            return reader.ReadUInt16();
        }
        public virtual int ReadInt32() {
            return reader.ReadInt32();
        }
        public virtual uint ReadUInt32() {
            return reader.ReadUInt32();
        }
        public virtual long ReadInt64() {
            return reader.ReadInt64();
        }
        public virtual ulong ReadUInt64() {
            return reader.ReadUInt64();
        }
        public virtual float ReadFloat() {
            return reader.ReadSingle();
        }
        public virtual double ReadDouble() {
            return reader.ReadDouble();
        }
        public virtual string ReadString() {
            var length = ReadUInt16();
            if (length == 0) return "";
            return Encoding.UTF8.GetString(reader.ReadBytes(length));
        }
        public virtual string ReadL10n(string key) {
            ReadString();
            return key;
        }
        public virtual DateTime ReadDateTime() {
            return BaseTime.AddMilliseconds(reader.ReadInt64());
        }
        public virtual byte[] ReadBytes() {
            var length = ReadInt32();
            return reader.ReadBytes(length);
        }
        public void Close() {
            reader.Dispose();
            if (closeStream) {
                stream.Dispose();
            }
        }
        public void Dispose() {
            Close();
        }
    }
}
